### PR TITLE
Detect revoked device and expired account

### PR DESCRIPTION
**Pull request by sajacl**
When app tries to recover connection,
Counting number of failed attempts, and after an interval, trying to diagnostic the error:

1. Check if device is revoked,
2. Check if account ran out of time
3. Report error

- - -
This change is [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)

### CODE REVIEW

**Review by pronebird — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)* status: 0 of 6 files reviewed, 7 unresolved discussions (waiting on @sajacl)

___
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 359 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0vXnvDueQrYK1PTum:-NG0vXnvDueQrYK1PTun:b-d205sf) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/PacketTunnel/PacketTunnelProvider.swift#L359)):*
> ```Swift
> 
>         guard numberOfFailedAttempts.isMultiple(of: 2) else {
>             startDiagnosticConnectionRecoveryFailingReason(
> ```

As it was outlined in spec, the tunnel should continue trying to reconnect, so performing those API requests shouldn't impede this.
___
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 492 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0wZSTBbfJYKioFSQ1:-NG0wZSTBbfJYKioFSQ2:b-v6usac) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/PacketTunnel/PacketTunnelProvider.swift#L492)):*
> ```Swift
>     // MARK: - Connection Recovery Diagnostic
> 
>     private enum ConnectionRecoveryDiagnosticErrors: Error {
> ```

In our case we can have multiple conditions met, i.e device revoked _and_ account expiry. I think it would be much simpler to log errors on spot and assign internal variables directly once both REST responses are received.
___
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 504 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0ucmF0WKYAnqjjgMl:-NG0ucmF0WKYAnqjjgMm:b-2e2728) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/PacketTunnel/PacketTunnelProvider.swift#L504)):*
> ```Swift
>         providerLogger.debug("Failed to recover connection, started diagnostic process")
> 
>         guard let deviceState = try? SettingsManager.readDeviceState(),
> ```

Error to read device state should be logged with:

```
providerLog.error(error: <error>, message: "<Describe when error occurred>")
```
___
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 548 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0tdew6q9lIkDFjOGv:-NG0tdew6q9lIkDFjOGw:b-vaszh5) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/PacketTunnel/PacketTunnelProvider.swift#L548)):*
> ```Swift
> 
>             if let accountData = accountCompletion?.value {
>                 if accountData.expiry > Date() {
> ```

The condition for account expiry seems to be inverse.
___
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 549 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0tPpu1ETu0D66WNgZ:-NG0tPpu1ETu0D66WNg_:b-nzcsbr) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/PacketTunnel/PacketTunnelProvider.swift#L549)):*
> ```Swift
>             if let accountData = accountCompletion?.value {
>                 if accountData.expiry > Date() {
>                     completion(.ranOutOfTime(newExpiry: accountData.expiry))
> ```

Completion handler will be called twice if both conditions (device revoked + account expiry) are met which would trigger double reconnect for the tunnel.
___
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 621 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0wAqH4jbss0zkcgvZ:-NG0wAqH4jbss0zkcgv_:bakl6pv) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/PacketTunnel/PacketTunnelProvider.swift#L621)):*
> ```Swift
>             providerLogger
>                 .debug(
>                     "Unable to fetch account data from network request, original error: \(error.localizedDescription)"
> ```

We have a helper to log errors:

```
providerLog.error(error: <error>, message: "<higher level description of when that error happened>")
```
___
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 638 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0uCrhAlox5TH44s0E:-NG0uCrhAlox5TH44s0F:b-uceltf) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/PacketTunnel/PacketTunnelProvider.swift#L638)):*
> ```Swift
>             accountExpiry = newExpiryDate
>             // Stop monitoring tunnel
>             tunnelMonitor.stop()
> ```

As per spec, tunnel monitor should stop when account is revoked too.

**Review by pronebird — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)* status: 0 of 6 files reviewed, 8 unresolved discussions (waiting on @sajacl)

___
*[`ios/MullvadREST/RESTRetryStrategy.swift` line 12 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0x__-4Aiigtb1CL0Z:-NG0x__-4Aiigtb1CL0_:b14wu5c) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/MullvadREST/RESTRetryStrategy.swift#L12)):*
> ```Swift
> 
> extension REST {
>     struct ExponentialBackoff: IteratorProtocol {
> ```

I think I have a better idea how to implement that and so I suggest we leave it out of this PR for simplicity and focus our efforts on the primary task.

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)* status: 0 of 6 files reviewed, 8 unresolved discussions (waiting on @sajacl)

___
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 638 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0uCrhAlox5TH44s0E:-NG0y-OR44GbKfp75Ys3:bs3qdwa) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/PacketTunnel/PacketTunnelProvider.swift#L638)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

As per spec, tunnel monitor should stop when account is revoked too.
</blockquote></details>

Typo: meant device revoked not account.

**Review by pronebird — CHANGES_REQUESTED** [review]
> Reviewed 3 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)* status: 3 of 6 files reviewed, 9 unresolved discussions (waiting on @sajacl)

___
*[`ios/MullvadVPN/TunnelManager/TunnelManager.swift` line 668 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0yRKc1JdrfgWPd3Rv:-NG0yRKc1JdrfgWPd3Rw:bgityuv) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/MullvadVPN/TunnelManager/TunnelManager.swift#L668)):*
> ```Swift
>         _tunnelStatus = newTunnelStatus
> 
>         if newTunnelStatus.packetTunnelStatus.isDeviceRevoked {
> ```

Note that we need a way to determine which events were already handled so that we do not attempt to revoke device or store data into keychain on each tunnel status update as it happens quite often.

**Comment by sajacl** [comment]
*[`ios/MullvadREST/RESTRetryStrategy.swift` line 12 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0x__-4Aiigtb1CL0Z:-NGGfDR_CfYMkvVj6Ny2:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/MullvadREST/RESTRetryStrategy.swift#L12)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

I think I have a better idea how to implement that and so I suggest we leave it out of this PR for simplicity and focus our efforts on the primary task.
</blockquote></details>

Done.

**Comment by sajacl** [comment]
*[`ios/MullvadVPN/TunnelManager/TunnelManager.swift` line 668 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0yRKc1JdrfgWPd3Rv:-NGGflht6KmzulEBpn4H:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/MullvadVPN/TunnelManager/TunnelManager.swift#L668)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Note that we need a way to determine which events were already handled so that we do not attempt to revoke device or store data into keychain on each tunnel status update as it happens quite often.
</blockquote></details>

Done.

**Comment by sajacl** [comment]
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 359 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0vXnvDueQrYK1PTum:-NGGfp6t85M9vmEqzZux:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/PacketTunnel/PacketTunnelProvider.swift#L359)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

As it was outlined in spec, the tunnel should continue trying to reconnect, so performing those API requests shouldn't impede this.
</blockquote></details>

Done.

**Comment by sajacl** [comment]
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 492 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0wZSTBbfJYKioFSQ1:-NGGftABELP_tsO_-iRt:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/PacketTunnel/PacketTunnelProvider.swift#L492)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

In our case we can have multiple conditions met, i.e device revoked _and_ account expiry. I think it would be much simpler to log errors on spot and assign internal variables directly once both REST responses are received.
</blockquote></details>

Done.

**Comment by sajacl** [comment]
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 504 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0ucmF0WKYAnqjjgMl:-NGGfuH51WOxmIdjCFfl:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/PacketTunnel/PacketTunnelProvider.swift#L504)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Error to read device state should be logged with:

```
providerLog.error(error: <error>, message: "<Describe when error occurred>")
```
</blockquote></details>

Done.

**Comment by sajacl** [comment]
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 548 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0tdew6q9lIkDFjOGv:-NGGfvhwElGgJQl9Z1eH:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/PacketTunnel/PacketTunnelProvider.swift#L548)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

The condition for account expiry seems to be inverse.
</blockquote></details>

Done.

**Comment by sajacl** [comment]
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 549 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0tPpu1ETu0D66WNgZ:-NGGfx9S-VgRh6nZOMil:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/PacketTunnel/PacketTunnelProvider.swift#L549)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Completion handler will be called twice if both conditions (device revoked + account expiry) are met which would trigger double reconnect for the tunnel.
</blockquote></details>

Done.

**Review by pronebird — CHANGES_REQUESTED** [review]
> Reviewed 3 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)* status: 4 of 6 files reviewed, 9 unresolved discussions (waiting on @sajacl)

___
*[`ios/MullvadTypes/PacketTunnelStatus.swift` line 32 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NGGfpLn1AbyvHHXJzVL:-NGGfpLn1AbyvHHXJzVM:bea6n53) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/cbe624139ce63db435e5b1fc499b133fd6402ad7/ios/MullvadTypes/PacketTunnelStatus.swift#L32)):*
> ```Swift
>     public var tunnelRelay: PacketTunnelRelay?
> 
>     public init(
> ```

We could perhaps leave single initializer that accepts `DeviceCheck?`

**Comment by sajacl** [comment]
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 621 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0wAqH4jbss0zkcgvZ:-NGGg1FWBZsxKpKf2T0_:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/PacketTunnel/PacketTunnelProvider.swift#L621)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

We have a helper to log errors:

```
providerLog.error(error: <error>, message: "<higher level description of when that error happened>")
```
</blockquote></details>

Done.

**Review by pronebird — CHANGES_REQUESTED** [review]
> Reviewed 1 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)* status: 5 of 6 files reviewed, 4 unresolved discussions (waiting on @sajacl)

___
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 356 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NGGoEuZ5RFPxRxv44tJ:-NGGoEuZ5RFPxRxv44tK:b-yw1ec3) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/cbe624139ce63db435e5b1fc499b133fd6402ad7/ios/PacketTunnel/PacketTunnelProvider.swift#L356)):*
> ```Swift
>         checkDeviceStateTask = nil
> 
>         isDeviceRevoked = false
> ```

You could store these fields along with `DeviceCheck` struct that you introduce and then simply assign `nil` to it.
___
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 368 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NGGgrLP8sFmNndG6uyr:-NGGgrLP8sFmNndG6uys:b-9ndrz6) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/cbe624139ce63db435e5b1fc499b133fd6402ad7/ios/PacketTunnel/PacketTunnelProvider.swift#L368)):*
> ```Swift
>         dispatchPrecondition(condition: .onQueue(dispatchQueue))
> 
>         if numberOfFailedAttempts.isMultiple(of: 2) {
> ```

`0.isMultiple(of: 2)` is also `true`

**Review by pronebird — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)* status: 5 of 6 files reviewed, 4 unresolved discussions (waiting on @sajacl)

___
*[`ios/MullvadVPN/TunnelManager/TunnelManager.swift` line 668 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NG0yRKc1JdrfgWPd3Rv:-NGGpCK70UtcwsrO4Db_:btcfj44) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcbae4bca8cdd6e3bae8a16010999cf8469501d2/ios/MullvadVPN/TunnelManager/TunnelManager.swift#L668)):*
<details><summary><i>Previously, sajacl (Sajad Vishkai) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Still not done. you'll be handling `DeviceCheck` on each call to `setTunenlStatus()`.

**Review by pronebird — COMMENTED** [review]
> Reviewed 1 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)* status: all files reviewed, 4 unresolved discussions (waiting on @sajacl)

**Review by pronebird — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)* status: all files reviewed, 6 unresolved discussions (waiting on @sajacl)

___
*[`ios/MullvadTypes/PacketTunnelStatus.swift` line 12 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NGH_fri5TixjfTVKFjy:-NGH_fri5TixjfTVKFjz:b-zi9a7c) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2f89bbcc24ea42d6571ad66b4a1c4266389a432a/ios/MullvadTypes/PacketTunnelStatus.swift#L12)):*
> ```Swift
> 
> public struct DeviceCheck: Codable, Equatable {
>     /// Flag indicating last changed date of device/account information changed from tunnel provider
> ```

Maybe we can reword it to:
```
/// Unique identifier for the device check. 
/// Should only change when other fields in the struct are being changed.
```
___
*[`ios/MullvadTypes/PacketTunnelStatus.swift` line 17 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NGH_kTw1CM5P1LHrCmu:-NGH_kTw1CM5P1LHrCmv:bslx4do) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2f89bbcc24ea42d6571ad66b4a1c4266389a432a/ios/MullvadTypes/PacketTunnelStatus.swift#L17)):*
> ```Swift
> 
>     /// Flag indicating device is revoked or not.
>     public var isDeviceRevoked: Bool?
> ```

Maybe we can reword it to:

```
/// Flag indicating whether device is revoked.
/// Set to `nil` when the device status is unknown yet.
```
___
*[`ios/MullvadTypes/PacketTunnelStatus.swift` line 20 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NGH_snOEqYA3e0t0ljL:-NGH_snOEqYA3e0t0ljM:b-i3s3d9) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2f89bbcc24ea42d6571ad66b4a1c4266389a432a/ios/MullvadTypes/PacketTunnelStatus.swift#L20)):*
> ```Swift
> 
>     /// Flag indicating that account expiry should be set again.
>     public var accountExpiry: Date?
> ```

Maybe we can reword it to:

```
/// Last known account expiry.
/// Set to `nil` when account expiry is unknown yet.
```
___
*[`ios/MullvadTypes/PacketTunnelStatus.swift` line 41 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NGHa6KA7gixhnwU_K8_:-NGHa6KB1LXviRsRVdNC:b-6l9upz) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2f89bbcc24ea42d6571ad66b4a1c4266389a432a/ios/MullvadTypes/PacketTunnelStatus.swift#L41)):*
> ```Swift
>     public var isNetworkReachable: Bool
> 
>     public var deviceCheck: DeviceCheck?
> ```

Maybe document it as 

```
/// Last performed device checked.
```

**Review by pronebird — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)* status: all files reviewed, 6 unresolved discussions (waiting on @sajacl)

___
*[`ios/MullvadVPN/TunnelManager/TunnelManager.swift` line 934 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NGHahSV8K7kShop4SCM:-NGHahSW3U25CsV8ShtF:bbf5o2f) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2f89bbcc24ea42d6571ad66b4a1c4266389a432a/ios/MullvadVPN/TunnelManager/TunnelManager.swift#L934)):*
> ```Swift
>             modificationBlock(&deviceState)
> 
>             if deviceState != oldState {
> ```

Maybe we don't need to filter out updates here as we should not have them running too often.

**Review by pronebird — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)* status: all files reviewed, 7 unresolved discussions (waiting on @sajacl)

___
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 224 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NGHbDtA0ijdqQJkMGux:-NGHbDtA0ijdqQJkMGuy:b9tlbb9) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2f89bbcc24ea42d6571ad66b4a1c4266389a432a/ios/PacketTunnel/PacketTunnelProvider.swift#L224)):*
> ```Swift
> 
>         dispatchQueue.async {
>             self.tunnelMonitor.stop()
> ```

Perhaps cancel `checkDeviceStateTask` in here as well.

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)* status: all files reviewed, 7 unresolved discussions (waiting on @sajacl)

___
*[`ios/MullvadTypes/PacketTunnelStatus.swift` line 17 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NGH_kTw1CM5P1LHrCmu:-NGHbbRi8zu1i-XCX-NG:b-p63wl2) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2f89bbcc24ea42d6571ad66b4a1c4266389a432a/ios/MullvadTypes/PacketTunnelStatus.swift#L17)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Maybe we can reword it to:

```
/// Flag indicating whether device is revoked.
/// Set to `nil` when the device status is unknown yet.
```
</blockquote></details>

Typo, should say: Set to `nil` when ~the~ device status is unknown yet.

**Review by pronebird — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)* status: 3 of 6 files reviewed, 6 unresolved discussions (waiting on @sajacl)

___
*[`ios/MullvadTypes/PacketTunnelStatus.swift` line 41 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NGHa6KA7gixhnwU_K8_:-NGHgjBG7gxuYdAbHc7C:b-lif4ou) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2f89bbcc24ea42d6571ad66b4a1c4266389a432a/ios/MullvadTypes/PacketTunnelStatus.swift#L41)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Maybe document it as 

```
/// Last performed device checked.
```
</blockquote></details>

Sorry I made a typo:

```
/// Last performed device check.
```

**Review by pronebird — COMMENTED** [review]
> Reviewed 3 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)* status: 4 of 6 files reviewed, 6 unresolved discussions (waiting on @sajacl)

**Comment by sajacl** [comment]
*[`ios/MullvadTypes/PacketTunnelStatus.swift` line 20 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NGH_snOEqYA3e0t0ljL:-NGHlwrI5zAj-9rw_nY0:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2f89bbcc24ea42d6571ad66b4a1c4266389a432a/ios/MullvadTypes/PacketTunnelStatus.swift#L20)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Maybe we can reword it to:

```
/// Last known account expiry.
/// Set to `nil` when account expiry is unknown yet.
```
</blockquote></details>

Done.

**Comment by sajacl** [comment]
*[`ios/MullvadTypes/PacketTunnelStatus.swift` line 41 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NGHa6KA7gixhnwU_K8_:-NGHlzJF9oygH-mE_jAd:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2f89bbcc24ea42d6571ad66b4a1c4266389a432a/ios/MullvadTypes/PacketTunnelStatus.swift#L41)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Sorry I made a typo:

```
/// Last performed device check.
```
</blockquote></details>

Done.

**Comment by sajacl** [comment]
*[`ios/MullvadTypes/PacketTunnelStatus.swift` line 32 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NGGfpLn1AbyvHHXJzVL:-NGHm-mMEcm-YYEN5jrK:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/cbe624139ce63db435e5b1fc499b133fd6402ad7/ios/MullvadTypes/PacketTunnelStatus.swift#L32)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

We could perhaps leave single initializer that accepts `DeviceCheck?`
</blockquote></details>

Done.

**Comment by sajacl** [comment]
*[`ios/MullvadVPN/TunnelManager/TunnelManager.swift` line 934 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NGHahSV8K7kShop4SCM:-NGHmeAd3FFT9HX5qFeV:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2f89bbcc24ea42d6571ad66b4a1c4266389a432a/ios/MullvadVPN/TunnelManager/TunnelManager.swift#L934)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Maybe we don't need to filter out updates here as we should not have them running too often.
</blockquote></details>

Done.

**Comment by sajacl** [comment]
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 224 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NGHbDtA0ijdqQJkMGux:-NGHn8xDCoHVqxCa2Nez:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2f89bbcc24ea42d6571ad66b4a1c4266389a432a/ios/PacketTunnel/PacketTunnelProvider.swift#L224)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Perhaps cancel `checkDeviceStateTask` in here as well.
</blockquote></details>

Done.



_Code quote:_
```Swift
        dispatchQueue.async {
            self.tunnelMonitor.stop()
```

**Comment by sajacl** [comment]
*[`ios/PacketTunnel/PacketTunnelProvider.swift` line 368 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NGGgrLP8sFmNndG6uyr:-NGHnQ2Q71WL_vyOWLAq:b-omhs4) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/cbe624139ce63db435e5b1fc499b133fd6402ad7/ios/PacketTunnel/PacketTunnelProvider.swift#L368)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

`0.isMultiple(of: 2)` is also `true`
</blockquote></details>

Value is increasing before the if statement.
Always starts from 1.

**Review by pronebird — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r9, 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)* status: all files reviewed, 1 unresolved discussion (waiting on @sajacl)

___
*[`ios/MullvadVPN/TunnelManager/TunnelManager.swift` line 927 at r10](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124#-NGHuRtd4ZlmOaQigLpR:-NGHuRtd4ZlmOaQigLpS:bt86nkd) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/302b315103af3a965102c7d0250ac84a27de6494/ios/MullvadVPN/TunnelManager/TunnelManager.swift#L927)):*
> ```Swift
>     ) {
>         let operation = AsyncBlockOperation(dispatchQueue: internalQueue) {
>             let oldState = self.deviceState
> ```

Can we please discard this change since 

```
let oldState = self.deviceState
var deviceState = oldState
```

is equivalent to what we had:

```
var deviceState = self.deviceState
```

**Review by pronebird — APPROVED** [review]
> Reviewed 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by pronebird — APPROVED** [review]
> Reviewed 1 of 1 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by pronebird — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4124)* status: :shipit: complete! all files reviewed, all discussions resolved
